File: iOS/Cirrent_OpenSource/Cirrent/SampleCloudService/SampleCloudService.swift
//
//  SampleCloudService.swift
//  Cirrent
//
//  Created by PSIHPOK on 1/9/17.
//  Copyright © 2017 PSIHPOK. All rights reserved.
//


public class SampleDevice {
    var deviceID:String!
    var friendlyName:String!
    var name:String!
    var imageURL:String!
}

@objc public enum LOGIN_RESULT_ : Int {
    case SUCCESS
    case FAILED_NO_RESPONSE
    case FAILED_INVALID_STATUS
}

@objc public enum BOUND_DEVICES_RESULT_ : Int {
    case SUCCESS
    case FAILED_NO_DEVICE
    case FAILED_NO_RESPONSE
    case FAILED_INVALID_STATUS
}

@objc public enum COMPLETION_RESULT_ : Int {
    case SUCCESS
    case NOT_LOGGED_IN
    case INVALID_DEVICE_ID
    case ERROR
}

public typealias LOGIN_RESULT = LOGIN_RESULT_
public typealias BOUND_DEVICES_RESULT = BOUND_DEVICES_RESULT_
public typealias COMPLETION_RESULT = COMPLETION_RESULT_

public typealias LoginCompletionHandler = (LOGIN_RESULT) -> Void
public typealias GetBoundDevicesCompletionHandler = (BOUND_DEVICES_RESULT, [SampleDevice]?) -> Void
public typealias BindCompletionHandler = (String?, COMPLETION_RESULT) -> Void
public typealias ResetCompletionHandler = (String?, COMPLETION_RESULT) -> Void
public typealias GetTokenCompletionHandler = (String?, COMPLETION_RESULT) -> Void

class SampleCloudService {

    static let sharedService:SampleCloudService = SampleCloudService()
    
    private let tokenURL = "https://go.cirrent.com/cloud/token/search"
    private let apiURL = "https://go.cirrent.com/cloud/"
    private let loginURL = "https://go.cirrent.com/api/login"
    
    private static let LOGGED_KEY = "logged_before_key"
    private static let OWNERID_KEY = "owner_id_key"
    private static let ID_ACCOUNT_KEY = "idaccount_key"
    private static let USERNAME_KEY = "username_key"
    private static let PASSWORD_KEY = "password_key"
    
    static let MANAGE_TOKEN_SCOPE = "manage"
    static let SEARCH_TOKEN_SCOPE = "search"
    static let BIND_TOKEN_SCOPE = "bind"
    
    var username:String? {
        get {
            let name = UserDefaults.standard.string(forKey: SampleCloudService.USERNAME_KEY)
            return name
        }
        
        set (name) {
            UserDefaults.standard.set(name, forKey: SampleCloudService.USERNAME_KEY)
        }
    }
    
    var password:String? {
        get {
            let pass = UserDefaults.standard.string(forKey: SampleCloudService.PASSWORD_KEY)
            return pass
        }
        
        set (pass) {
            UserDefaults.standard.set(pass, forKey: SampleCloudService.PASSWORD_KEY)
        }
    }
    
    var bLogged:Bool {
        get {
            let bLogged = UserDefaults.standard.bool(forKey: SampleCloudService.LOGGED_KEY)
            return bLogged
        }
        
        set(logged) {
            UserDefaults.standard.set(logged, forKey: SampleCloudService.LOGGED_KEY)
        }
    }
    
    var ownerID:String? {
        get {
            let ownerID = UserDefaults.standard.string(forKey: SampleCloudService.OWNERID_KEY)
            return ownerID
        }
        
        set(id) {
            UserDefaults.standard.set(id, forKey: SampleCloudService.OWNERID_KEY)
        }
    }
    
    var accountID:String? {
        get {
            let accountID = UserDefaults.standard.string(forKey: SampleCloudService.ID_ACCOUNT_KEY)
            return accountID
        }
        
        set(id) {
            UserDefaults.standard.set(id, forKey: SampleCloudService.ID_ACCOUNT_KEY)
        }
    }
    
    var bindedDevice:Bool = false
    
    var manage_token:String? = nil
    var search_token:String? = nil
    var bind_token:String? = nil
    
    init() {
        
    }
    
    func login(username:String, password:String, completion: @escaping LoginCompletionHandler) {
        self.username = username
        self.password = password
        
        let encoded_username = username.addingPercentEncoding(withAllowedCharacters: .urlHostAllowed)
        let encoded_password = password.addingPercentEncoding(withAllowedCharacters: .urlHostAllowed)
        
        let url = "\(loginURL)?username=\(encoded_username!)&password=\(encoded_password!)"
        
        sendPostRequest(url: url, completionHandler: {
            data, response, error in
            
            guard let data = data, error == nil else {
                LogService.sharedService.debug(data: "LOGIN-FAILED - CLOUD CONNECTION ERROR")
                completion(.FAILED_NO_RESPONSE)
                return
            }
            
            if let httpStatus = response as? HTTPURLResponse, httpStatus.statusCode != 200 {
                LogService.sharedService.debug(data: "LOGIN-FAILED - INVALID_STATUS:\(httpStatus.statusCode)")
                completion(.FAILED_INVALID_STATUS)
                return
            }
            
            let json = JSON(data: data)
            if json == nil {
                LogService.sharedService.debug(data: "LOGIN-FAILED - INVALID RESPONSE, JSON PARSE ERROR")
                completion(.FAILED_NO_RESPONSE)
                return
            }
            
            let user = json["data"]["user"]
            if user == nil {
                LogService.sharedService.debug(data: "LOGIN-FAILED - INVALID RESPONSE, JSON PARSE ERROR")
                completion(.FAILED_NO_RESPONSE)
                return
            }
            
            if user["idAccount"] == nil || user["idAccount"].stringValue == "" {
                LogService.sharedService.debug(data: "LOGIN-FAILED - INVALID ACCOUNT ID")
                completion(.FAILED_NO_RESPONSE)
                return
            }
            
            self.accountID = user["idAccount"].stringValue
            self.ownerID = username
            LogService.sharedService.debug(data: "LOGIN-SUCCESS")
            completion(.SUCCESS)
        })
    }
    
    func logOut() {
        username = nil
        password = nil
        ownerID = nil
        accountID = nil
        bLogged = false
    }
    
    func getBoundDevices(completion: @escaping GetBoundDevicesCompletionHandler) {
        let url = apiURL + "/devices"
        sendGetRequest(url: url, completionHandler: {
            data, response, error in
            
            guard let _ = data, error == nil else {
                LogService.sharedService.debug(data: "SampleCloud - Offline")
                completion(BOUND_DEVICES_RESULT.FAILED_NO_RESPONSE, nil)
                return
            }
            
            if let httpStatus = response as? HTTPURLResponse, httpStatus.statusCode != 200 {
                LogService.sharedService.debug(data: "SampleCloud - Get Devices From Account - INVALID_STATUS:\(httpStatus.statusCode)")
                completion(BOUND_DEVICES_RESULT.FAILED_INVALID_STATUS, nil)
                return
            }
            
            let str = String(data: data!, encoding: .utf8)
            print(str!)
            
            let jsonData:JSON = JSON(data: data!)
            self.manage_token = jsonData["manage_token"].stringValue
            let jsonArray = jsonData["devices"]
            
            var deviceArray:[SampleDevice] = [SampleDevice]()
            for item in jsonArray {
                let device = self.getSampleDeviceFromJson(data: item.1)
                deviceArray.append(device)
            }
            
            if deviceArray.count == 0 {
                LogService.sharedService.debug(data: "SampleCloud - Get Devices From Account - NO_DEVICES")
                completion(BOUND_DEVICES_RESULT.FAILED_NO_DEVICE, nil)
            }
            else {
                LogService.sharedService.debug(data: "SampleCloud - Get Devices From Account Success")
                completion(BOUND_DEVICES_RESULT.SUCCESS, deviceArray)
            }
        })
    }
    
    public func getToken(tokenType:TOKEN_TYPE, deviceID:String?, completion:@escaping (_ token:String?) -> Void) {
        if tokenType == .SEARCH {
            if search_token == nil {
                getSearchToken(completion: {
                    token, response in
                    
                    if response != .SUCCESS {
                        completion(nil)
                    }
                    else {
                        completion(self.search_token)
                    }
                })
            }
            else {
                completion(self.search_token)
            }
        }
        else if tokenType == .BIND {
            if bind_token == nil && deviceID != nil {
                bindDevice(deviceID: deviceID!, friendlyName: nil, completion: {
                    token, response in
                    
                    if response != .SUCCESS {
                        completion(nil)
                    }
                    else {
                        completion(self.bind_token)
                    }
                })
            }
            else {
                completion(self.bind_token)
            }
        }
        else if tokenType == .MANAGE {
                self.getBoundDevices(completion: {
                    response, deviceArray in
                    
                    completion(self.manage_token)
                })
        }
        else {
            if search_token != nil {completion(search_token)}
            else if bind_token != nil {completion(bind_token)}
            else {completion(manage_token)}
        }
    }
    
    public func bindDevice(deviceID:String, friendlyName:String?, completion: @escaping BindCompletionHandler) {
        let url = apiURL + "/bind/" + deviceID
        var params:[String:AnyObject] = [:]
        if friendlyName != nil && friendlyName! != "" {
            params = ["friendly_name" : friendlyName! as AnyObject]
        }
        
        sendPostRequest(url: url, parameters: params, completionHandler: {
            data, response, error in
            guard let _ = data, error == nil else {
                completion(nil, .ERROR)
                return
            }
            
            if let httpStatus = response as? HTTPURLResponse, httpStatus.statusCode != 200 {
                if httpStatus.statusCode == 401 {
                    LogService.sharedService.debug(data: "SampleCloud - Bind Device Failed - NOT_LOGGED_IN")
                    completion(nil, .NOT_LOGGED_IN)
                }
                else if httpStatus.statusCode == 404 {
                    LogService.sharedService.debug(data: "SampleCloud - Bind Device Failed - INVALID_DEVICE_ID")
                    completion(nil, .INVALID_DEVICE_ID)
                }
                else {
                    LogService.sharedService.debug(data: "SampleCloud - Bind Device Failed - INVALID_STATUS:\(httpStatus.statusCode)")
                    completion(nil, .ERROR)
                }
                return
            }
            
            let jsonData = JSON(data: data!)
            
            self.bind_token = jsonData["token"].stringValue
            self.manage_token = jsonData["token"].stringValue
            var logStr = "Type=BIND;value=" + self.bind_token!
            LogService.sharedService.log(event: .TOKEN_RECEIVED, data: logStr)
            logStr = "Type=MANAGE;value=" + self.manage_token!
            LogService.sharedService.log(event: .TOKEN_RECEIVED, data: logStr)
            LogService.sharedService.debug(data: "SampleCloud - Binded Device - deviceID=\(deviceID)")
            completion(self.bind_token, .SUCCESS)
        })
    }
    
    public func resetDevice(deviceID:String, completion: @escaping ResetCompletionHandler) {
        let url = apiURL + "/reset/" + deviceID
        
        sendDeleteRequest(url: url, completionHandler: {
            data, response, error in
            
            guard let _ = data, error == nil else {
                LogService.sharedService.log(event: .CLOUD_CONNECTION_ERROR, data: "")
                completion(nil, .ERROR)
                return
            }
            
            if let httpStatus = response as? HTTPURLResponse, httpStatus.statusCode != 200 {
                if httpStatus.statusCode == 401 {
                    LogService.sharedService.debug(data: "SampleCloud - Reset Device Failed - INVALID_TOKEN")
                    completion(nil, .ERROR)
                }
                else if httpStatus.statusCode == 404 {
                    LogService.sharedService.debug(data: "SampleCloud - Reset Device Failed - INVALID_DEVICE_ID")
                    completion(nil, .INVALID_DEVICE_ID)
                }
                else {
                    LogService.sharedService.debug(data: "SampleCloud - Reset Device Failed - INVALID_STATUS:\(httpStatus.statusCode)")
                    completion(nil, .ERROR)
                }
                return
            }
            
            let jsonData = JSON(data: data!)
            self.manage_token = jsonData["token"].stringValue
            let logStr = "Type=MANAGE;value=" + self.manage_token!
            LogService.sharedService.log(event: .TOKEN_RECEIVED, data: logStr)
            LogService.sharedService.debug(data: "SampleCloud - Reset Device - deviceID=\(deviceID)")
            completion(self.manage_token, .SUCCESS)
        })
    }
    
    private func getSearchToken(completion: @escaping GetTokenCompletionHandler) {
        sendGetRequest(url: tokenURL, completionHandler: {
            data, response, error in
            guard let data = data, error == nil else {
                LogService.sharedService.log(event: .TOKEN_ERROR, data: "Error=NO_RESPONSE")
                completion(nil, .ERROR)
                return
            }
            
            if let httpStatus = response as? HTTPURLResponse, httpStatus.statusCode != 200 {
                if httpStatus.statusCode == 401 {
                    LogService.sharedService.debug(data: "SampleCloud - Get Search Token Failed - NOT_LOGGED_IN")
                    completion(nil, .NOT_LOGGED_IN)
                }
                else {
                    LogService.sharedService.debug(data: "SampleCloud - Get Search Token Failed - INVALID_STATUS:\(httpStatus.statusCode)")
                    completion(nil, .ERROR)
                }
                return
            }
            
            let json = JSON(data: data)
            let token:String = String(describing: json["token"])
            self.search_token = token
            
            let logStr = "Type=SEARCH;value=" + token
            LogService.sharedService.log(event: .TOKEN_RECEIVED, data: logStr)
            completion(self.search_token, .SUCCESS)
        })
    }
    
    private func getSampleDeviceFromJson(data:JSON) -> SampleDevice {
        let device = SampleDevice()
        
        device.deviceID = data["cirrent_device_id"].stringValue
        device.friendlyName = data["friendly_name"].stringValue
        device.name = data["device_type_name"].stringValue
        device.imageURL = data["device_type_image"].stringValue
        
        return device
    }
    
    private func sendGetRequest(url: String, completionHandler: @escaping (Data?, URLResponse?, Error?) -> Void) {
        let requestURL = URL(string:url)!
        var request = URLRequest(url: requestURL)
        request.httpMethod = "GET"
        request.setValue("Basic \(getAuthenticationString())", forHTTPHeaderField: "Authorization")
        
        let task = URLSession.shared.dataTask(with: request, completionHandler: completionHandler)
        task.resume()
    }
    
    private func sendPostRequest(url: String, parameters: [String: AnyObject], completionHandler: @escaping (Data?, URLResponse?, Error?) -> Void) {
        let parameterString = parameters.stringFromHttpParameters()
        var request = URLRequest(url: URL(string: url)!)
        
        request.httpBody = parameterString.data(using: .utf8)
        request.httpMethod = "POST"
        request.setValue("Basic \(getAuthenticationString())", forHTTPHeaderField: "Authorization")
        
        let task = URLSession.shared.dataTask(with: request, completionHandler: completionHandler)
        task.resume()
    }
    
    private func sendPostRequest(url: String, completionHandler: @escaping (Data?, URLResponse?, Error?) -> Void) {
        
        var request = URLRequest(url: URL(string: url)!)
        request.httpMethod = "POST"
        let task = URLSession.shared.dataTask(with: request, completionHandler: completionHandler)
        task.resume()
    }
    
    private func sendDeleteRequest(url:String, completionHandler: @escaping (Data?, URLResponse?, Error?) -> Void) {
        var request = URLRequest(url: URL(string: url)!)
        request.httpMethod = "DELETE"
        request.setValue("Basic \(getAuthenticationString())", forHTTPHeaderField: "Authorization")
        let task = URLSession.shared.dataTask(with: request, completionHandler: completionHandler)
        task.resume()
    }
    
    private func getAuthenticationString() -> String {
        if username == nil || password == nil {
            return ""
        }
        
        let loginString = String(format: "%@:%@", username!, password!)
        let loginData = loginString.data(using: String.Encoding.utf8)!
        let base64LoginString = loginData.base64EncodedString()
        return base64LoginString
    }
}

File: Android/Cirrent_Android/app/src/main/java/com/cirrent/cirrent/SampleCloudService.java
package com.cirrent.cirrent;

import android.content.Context;
import android.content.SharedPreferences;

import com.cirrent.cirrentsdk.LogService;
import com.loopj.android.http.AsyncHttpClient;
import com.loopj.android.http.JsonHttpResponseHandler;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

import cz.msebera.android.httpclient.entity.ContentType;
import cz.msebera.android.httpclient.entity.StringEntity;
import cz.msebera.android.httpclient.Header;

public class SampleCloudService {

    private static SampleCloudService sharedService = null;
    private Context mContext = null;
    private static AsyncHttpClient client = new AsyncHttpClient();

    private static String TOKEN_URL = "https://go.cirrent.com/cloud/token/search";
    private static String apiURL = "https://go.cirrent.com/cloud/";
    private static String PREF_KEY = "prefs";

    private static String USERNAME_KEY = "username";
    private static String PASSWORD_KEY = "password";

    private static final int TIME_OUT = 10000;
    private static final int CONNECTIONS = 1;
    String appID;
    String username;
    String password;

    String manage_token = null;
    String search_token = null;
    String bind_token = null;

    boolean bindedDevice = false;

    public SampleCloudService() {

    }

    public static SampleCloudService sharedService() {
        if (sharedService == null) {
            sharedService = new SampleCloudService();
            client.setTimeout(TIME_OUT);
            client.setConnectTimeout(TIME_OUT);
            client.setResponseTimeout(TIME_OUT);
            client.setMaxConnections(CONNECTIONS);
        }

        return sharedService;
    }

    public void setContext(Context context) {
        mContext = context;
    }

    void saveLogInInfo(String username, String password) {
        if (mContext == null) return;

        this.username = username;
        this.password = password;
        this.appID = username;

        SharedPreferences prefs = mContext.getSharedPreferences(PREF_KEY, 0);
        SharedPreferences.Editor editor = prefs.edit();

        editor.putString(USERNAME_KEY, username);
        editor.putString(PASSWORD_KEY, password);

        editor.commit();
    }

    String getUsername() {
        if (mContext == null) return null;

        SharedPreferences prefs = mContext.getSharedPreferences(PREF_KEY, 0);
        String username = prefs.getString(USERNAME_KEY, null);
        return username;
    }

    String getPassword() {
        if (mContext == null) return null;

        SharedPreferences prefs = mContext.getSharedPreferences(PREF_KEY, 0);
        String password = prefs.getString(PASSWORD_KEY, null);
        return password;
    }

    void login(final String username, final String password, final SampleCompletionHandler handler) {
        this.username = username;
        this.password = password;
        getSearchToken(new SampleCompletionHandler(){
            @Override
            public void completionGetSearchToken(boolean bSuccess) {
                if (bSuccess == true) {
                    saveLogInInfo(username, password);
                }
                handler.completion(bSuccess);
            }
        });
    }

    void signOut() {
        username = null;
        password = null;
        manage_token = null;
        search_token = null;
        bind_token = null;
        appID = null;

        SharedPreferences prefs = mContext.getSharedPreferences(PREF_KEY, 0);
        SharedPreferences.Editor editor = prefs.edit();

        editor.remove(USERNAME_KEY);
        editor.remove(PASSWORD_KEY);

        editor.commit();
    }

    void getSearchToken(final SampleCompletionHandler handler) {
        getRequest(TOKEN_URL, new JSONObject(), new JsonHttpResponseHandler() {
            @Override
            public void onSuccess(int statusCode, Header[] headers, JSONObject response) {
                try {
                    String token = response.getString("token");
                    LogService.sharedService().log(LogService.Event.TOKEN_RECEIVED, "Type=SEARCH;value=" + token);
                    search_token = token;
                    handler.completionGetSearchToken(true);
                } catch (JSONException e) {
                    e.printStackTrace();
                    LogService.sharedService().log(LogService.Event.TOKEN_ERROR, "Error=PARSE_JSON_ERROR");
                    handler.completionGetSearchToken(false);
                }
            }

            @Override
            public void onFailure(int statusCode, Header[] headers, Throwable throwable, JSONObject errorResponse) {
                if (statusCode == 0) {
                    LogService.sharedService().log(LogService.Event.CLOUD_CONNECTION_ERROR, "");
                }
                else if (statusCode == 401) {
                    LogService.sharedService().log(LogService.Event.TOKEN_ERROR, "Error=Unauthorized - authentication credentials are invalid");
                }
                else if (statusCode == 404) {
                    LogService.sharedService().log(LogService.Event.TOKEN_ERROR, "Error=Device id is not recognized or is not associated with the user's account");
                }
                else {
                    LogService.sharedService().log(LogService.Event.TOKEN_ERROR, "Error=Invalid Status " + statusCode);
                }
                handler.completionGetSearchToken(false);
            }
        });
    }

    public class SampleDevice {
        String deviceID = null;
        String friendlyName = null;
        String name = null;
        String imageURL = null;
    }

    public static class SampleCompletionHandler {
        public void completionGetSearchToken(boolean bSuccess){}
        public void completionGetDevices(ArrayList<SampleDevice> devices) {}
        public void completionBindDevice(boolean bSuccess) {}
        public void completionResetDevice(boolean bSuccess){}
        public void completion(boolean bSuccess){}
    }

    public void getDevices(final SampleCompletionHandler handler) {
        String url = apiURL + "/devices";
        getRequest(url, new JSONObject(), new JsonHttpResponseHandler() {
            @Override
            public void onSuccess(int statusCode, Header[] headers, JSONObject response) {
                try {
                    manage_token = response.getString("manage_token");
                    JSONArray deviceArray = response.getJSONArray("devices");
                    ArrayList<SampleDevice> devices = new ArrayList<SampleDevice>();
                    for (int i = 0; i < deviceArray.length(); i++) {
                        JSONObject deviceData = deviceArray.getJSONObject(i);
                        SampleDevice device = getDeviceFromJson(deviceData);
                        devices.add(device);
                    }
                    handler.completionGetDevices(devices);
                }catch (JSONException e) {
                    e.printStackTrace();
                    LogService.sharedService().debug("SampleCloud - ");
                    handler.completionGetDevices(null);
                }
            }

            @Override
            public void onFailure(int statusCode, Header[] headers, Throwable throwable, JSONObject errorResponse) {
                if (statusCode == 0) {
                    LogService.sharedService().log(LogService.Event.CLOUD_CONNECTION_ERROR, "");
                }
                else if (statusCode == 401) {
                    LogService.sharedService().log(LogService.Event.TOKEN_ERROR, "Error=Unauthorized - authentication credentials are invalid");
                }
                else if (statusCode == 404) {
                    LogService.sharedService().log(LogService.Event.TOKEN_ERROR, "Error=Device id is not recognized or is not associated with the user's account");
                }
                else {
                    LogService.sharedService().log(LogService.Event.TOKEN_ERROR, "Error=Invalid Status " + statusCode);
                }
                LogService.sharedService().debug("SampleCloud - Get Devices Failed - Invalid Status " + statusCode);
                handler.completionGetDevices(null);
            }
        });
    }

    public void bindDevice(final String deviceID, String friendlyName, final SampleCompletionHandler handler) {
        String url = apiURL + "/bind/" + deviceID;
        JSONObject param = new JSONObject();
        try {
            if (friendlyName != null) {
                param.put("friendly_name", friendlyName);
            }
        }catch (JSONException e) {
            e.printStackTrace();
        }

        postRequest(url, param, new JsonHttpResponseHandler() {
            @Override
            public void onSuccess(int statusCode, Header[] headers, JSONObject response) {
                try {
                    bind_token = response.getString("token");
                }catch (JSONException e) {
                    e.printStackTrace();
                }

                String logStr = "Type=BIND;value=" + bind_token;
                LogService.sharedService().log(LogService.Event.TOKEN_RECEIVED, logStr);
                LogService.sharedService().debug("SampleCloud - Binded Device - DeviceID=" + deviceID);
                handler.completionBindDevice(true);
            }

            @Override
            public void onFailure(int statusCode, Header[] headers, Throwable throwable, JSONObject errorResponse) {
                if (statusCode == 0) {
                    LogService.sharedService().log(LogService.Event.CLOUD_CONNECTION_ERROR, "");
                }
                else if (statusCode == 401) {
                    LogService.sharedService().log(LogService.Event.TOKEN_ERROR, "Error=Unauthorized - authentication credentials are invalid");
                }
                else if (statusCode == 404) {
                    LogService.sharedService().log(LogService.Event.TOKEN_ERROR, "Error=Device id is not recognized or is not associated with the user's account");
                }
                else {
                    LogService.sharedService().log(LogService.Event.TOKEN_ERROR, "Error=Invalid Status " + statusCode);
                }
                LogService.sharedService().debug("SampleCloud - Bind Device Failed - Invalid Status " + statusCode);
                handler.completionBindDevice(false);
            }
        });
    }

    public void resetDevice(final String deviceID, final SampleCompletionHandler handler) {
        String url = apiURL + "/reset/" + deviceID;
        deleteRequest(url, new JsonHttpResponseHandler() {
            @Override
            public void onSuccess(int statusCode, Header[] headers, JSONObject response) {
                LogService.sharedService().debug("SampleCloud - Reset Device - DeviceID=" + deviceID);
                handler.completionResetDevice(true);
            }

            @Override
            public void onFailure(int statusCode, Header[] headers, Throwable throwable, JSONObject errorResponse) {
                String logStr = "SampleCloud - Reset Device Failed - DeviceID:" + deviceID + " ";
                if (statusCode == 0) {
                    LogService.sharedService().log(LogService.Event.CLOUD_CONNECTION_ERROR, "");
                    handler.completionResetDevice(false);
                    return;
                }
                else if (statusCode == 401) {
                    logStr += "Unauthorized - authentication credentials are invalid";
                }
                else if (statusCode == 404) {
                    logStr += "Device id is not recognized or is not associated with the user's account";
                }
                else {
                    logStr += "Invalid Status " + statusCode;
                }
                LogService.sharedService().debug(logStr);
                handler.completionResetDevice(false);
            }
        });
    }

    private SampleDevice getDeviceFromJson(JSONObject data) {
        SampleDevice device = new SampleDevice();
        try {device.deviceID = data.getString("cirrent_device_id");}catch (JSONException e) {e.printStackTrace();}
        try {device.friendlyName = data.getString("friendly_name");}catch (JSONException e) {e.printStackTrace();}
        try {device.name = data.getString("device_type_name");}catch (JSONException e) {e.printStackTrace();}
        try {device.imageURL = data.getString("device_type_image");}catch (JSONException e) {e.printStackTrace();}
        return device;
    }

    private void postRequest(String url, JSONObject object, JsonHttpResponseHandler handler) {
        StringEntity entity = new StringEntity(object.toString(), ContentType.APPLICATION_JSON);
        if (getUsername() == null || getPassword() == null) {
            if (this.username != null && this.password != null) {
                client.setBasicAuth(this.username, this.password);
            }
        }
        else {
            client.setBasicAuth(getUsername(), getPassword());
        }

        client.post(this.mContext, url, entity, ContentType.APPLICATION_JSON.getMimeType(), handler);
    }

    private void getRequest(String url, JSONObject object, JsonHttpResponseHandler handler) {
        StringEntity entity = new StringEntity(object.toString(), ContentType.APPLICATION_JSON);
        if (getUsername() == null || getPassword() == null) {
            if (this.username != null && this.password != null) {
                client.setBasicAuth(this.username, this.password);
            }
        }
        else {
            client.setBasicAuth(getUsername(), getPassword());
        }
        client.get(this.mContext, url, entity, ContentType.APPLICATION_JSON.getMimeType(), handler);
    }

    private void deleteRequest(String url, JsonHttpResponseHandler handler) {
        client.setBasicAuth(getUsername(), getPassword());
        client.delete(this.mContext, url, null, handler);
    }
}
